File: tools/pcd/batch_former.py
"""Round 2 batch formation from collapsed manifest and structural map.

Groups preserved triage cards into review batches by cluster, routes
cross-file questions to the batch containing target files, and attaches
cluster context for deep review.
"""

from __future__ import annotations

import os
from collections import defaultdict
from typing import Any


# ---------------------------------------------------------------------------
# Token estimation
# ---------------------------------------------------------------------------


def _estimate_tokens(files: list[str], project_root: str) -> int:
    """Rough token estimate for batch input (file content chars / 4).

    Falls back to line-count heuristic from triage card data when files
    are not readable on disk.
    """
    total_chars = 0
    for file_path in files:
        full_path = os.path.join(project_root, file_path)
        try:
            total_chars += os.path.getsize(full_path)
        except OSError:
            # Fallback: assume ~80 chars per line, ~200 lines
            total_chars += 16000
    return max(total_chars // 4, 1) if total_chars > 0 else 0


# ---------------------------------------------------------------------------
# Question routing
# ---------------------------------------------------------------------------


def _route_questions(
    question_index: list[dict[str, Any]],
    batches: list[dict[str, Any]],
) -> None:
    """Route questions from triage manifest to appropriate batches in-place.

    For each question, find the batch containing any of the target_files.
    If no batch contains the target, add to the batch with the most file
    overlap with the question's from_file directory.
    """
    if not batches:
        return

    for q in question_index:
        target_files: list[str] = q.get("target_files", [])
        from_file: str = q.get("from_file", "")
        q_text: str = q.get("text", q.get("question", ""))
        q_id: str = q.get("id", "")

        routed_question = {
            "id": q_id,
            "from_file": from_file,
            "question": q_text,
        }

        # Try direct match: batch contains a target file
        routed = False
        for batch in batches:
            batch_files = set(batch["files"])
            if batch_files & set(target_files):
                batch["routed_questions"].append(routed_question)
                routed = True
                break

        if routed:
            continue

        # Fallback: batch with most overlap by directory
        from_dir = os.path.dirname(from_file)
        best_batch = batches[0]
        best_overlap = -1
        for batch in batches:
            overlap = sum(
                1
                for f in batch["files"]
                if os.path.dirname(f) == from_dir
            )
            if overlap > best_overlap:
                best_overlap = overlap
                best_batch = batch

        best_batch["routed_questions"].append(routed_question)


# ---------------------------------------------------------------------------
# Cluster helpers
# ---------------------------------------------------------------------------


def _build_file_to_cluster(structural_map: dict[str, Any]) -> dict[str, int]:
    """Map each file path to its cluster_id from the structural map."""
    mapping: dict[str, int] = {}
    for module in structural_map.get("modules", []):
        path = module.get("path", "")
        cluster_id = module.get("cluster_id", 0)
        mapping[path] = cluster_id
    return mapping


def _get_cluster_by_id(
    structural_map: dict[str, Any],
    cluster_id: int,
) -> dict[str, Any] | None:
    """Find a cluster dict by its id in the structural map."""
    for cluster in structural_map.get("clusters", []):
        if cluster.get("id") == cluster_id:
            return cluster
    return None


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def form_batches(
    collapsed_manifest: dict[str, Any],
    structural_map: dict[str, Any],
    max_files_per_batch: int = 15,
    project_root: str = ".",
) -> list[dict[str, Any]]:
    """Form Round 2 review batches from collapsed manifest and structural map.

    Strategy:
    1. Group preserved cards by cluster_id from structural map
    2. If a cluster group exceeds max_files_per_batch, split by sub-directory
    3. Route questions to the batch containing the target file
    4. Attach relevant cluster context

    Returns:
        List of batch definition dicts conforming to BATCH_DEFINITION_SCHEMA.
    """
    preserved_cards: list[dict[str, Any]] = collapsed_manifest.get(
        "preserved_cards", []
    )
    question_index: list[dict[str, Any]] = collapsed_manifest.get(
        "question_index", []
    )

    if not preserved_cards:
        return []

    file_to_cluster = _build_file_to_cluster(structural_map)

    # Group cards by cluster_id
    cluster_groups: dict[int, list[dict[str, Any]]] = defaultdict(list)
    for card in preserved_cards:
        file_path = card.get("file", "")
        cluster_id = file_to_cluster.get(file_path, 0)
        cluster_groups[cluster_id].append(card)

    # Build batches
    raw_batches: list[dict[str, Any]] = []
    batch_id_counter = 1

    for cluster_id, cards in sorted(cluster_groups.items()):
        if len(cards) <= max_files_per_batch:
            files = [c["file"] for c in cards]
            cluster_ctx = _get_cluster_by_id(structural_map, cluster_id)
            raw_batches.append(
                {
                    "batch_id": batch_id_counter,
                    "files": files,
                    "triage_cards": cards,
                    "cluster_context": [cluster_ctx] if cluster_ctx else [],
                    "routed_questions": [],
                    "estimated_tokens": _estimate_tokens(files, project_root),
                }
            )
            batch_id_counter += 1
        else:
            # Split by sub-directory
            subdir_groups: dict[str, list[dict[str, Any]]] = defaultdict(list)
            for card in cards:
                subdir = os.path.dirname(card.get("file", ""))
                subdir_groups[subdir].append(card)

            cluster_ctx = _get_cluster_by_id(structural_map, cluster_id)

            for _subdir, sub_cards in sorted(subdir_groups.items()):
                # Further split if a single subdir still exceeds the limit
                for i in range(0, len(sub_cards), max_files_per_batch):
                    chunk = sub_cards[i : i + max_files_per_batch]
                    files = [c["file"] for c in chunk]
                    raw_batches.append(
                        {
                            "batch_id": batch_id_counter,
                            "files": files,
                            "triage_cards": chunk,
                            "cluster_context": (
                                [cluster_ctx] if cluster_ctx else []
                            ),
                            "routed_questions": [],
                            "estimated_tokens": _estimate_tokens(
                                files, project_root
                            ),
                        }
                    )
                    batch_id_counter += 1

    # Route questions to batches
    _route_questions(question_index, raw_batches)

    return raw_batches

File: packages/cli/src/topics/pcd/batch-former.ts
/**
 * PCD Round 2 batch formation.
 *
 * Ported 1:1 from `tools/pcd/batch_former.py`. Groups preserved triage cards
 * into review batches by structural-map cluster, splits oversized clusters by
 * sub-directory, and routes cross-file questions to the batch containing the
 * question's target file.
 *
 * Parity contract with the Python reference:
 *   - Cluster grouping is keyed on `cluster_id` looked up via the structural
 *     map's module list. Files missing from the map fall back to cluster 0.
 *   - Cluster groups are emitted in ascending `cluster_id` order (Python
 *     `sorted(cluster_groups.items())`).
 *   - Oversized clusters split alphabetically by `dirname(file)`; each subdir
 *     chunk is then sliced into `max_files_per_batch` runs.
 *   - `batch_id` is a 1-based integer counter assigned in emission order.
 *   - `estimated_tokens` uses on-disk file size / 4, with a 16000-char
 *     fallback per missing file. Empty input -> 0.
 *   - Questions route to the first batch whose `files` share ANY entry with
 *     the question's `target_files`. Fallback: batch with the most files in
 *     the same directory as `from_file` (first batch wins on tie).
 */

import { statSync } from 'node:fs';
import { join } from 'node:path';
import type { CollapsedManifest, TriageCard } from './collapse';
import type { StructuralMap, StructuralMapCluster } from './structural-map';

// ---------------------------------------------------------------------------
// Path helpers (Python-parity)
// ---------------------------------------------------------------------------

/**
 * Python-parity `os.path.dirname` for POSIX paths.
 *
 * Differs from Node's `path.dirname` in two cases that matter here:
 *   - `""` -> `""` (Node: `"."`).
 *   - `"foo.py"` -> `""` (Node: `"."`).
 *
 * Triage card file paths are always project-relative with `/` separators,
 * so a stdlib-style split on the last `/` matches Python byte-for-byte.
 */
function pyDirname(path: string): string {
  const idx = path.lastIndexOf('/');
  return idx >= 0 ? path.slice(0, idx) : '';
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface RoutedQuestion {
  id: string;
  from_file: string;
  question: string;
}

export interface BatchDefinition {
  batch_id: number;
  files: string[];
  triage_cards: TriageCard[];
  cluster_context: StructuralMapCluster[];
  routed_questions: RoutedQuestion[];
  estimated_tokens: number;
}

// ---------------------------------------------------------------------------
// Token estimation
// ---------------------------------------------------------------------------

/**
 * Rough token estimate for batch input: sum of file sizes / 4. When a file
 * cannot be stat'd (missing, permission error), falls back to 16000 chars
 * (~200 lines * 80 chars). Returns 0 iff the total char count is 0.
 */
export function _estimateTokens(files: string[], projectRoot: string): number {
  let totalChars = 0;
  for (const filePath of files) {
    const fullPath = join(projectRoot, filePath);
    try {
      totalChars += statSync(fullPath).size;
    } catch {
      totalChars += 16000;
    }
  }
  if (totalChars === 0) return 0;
  return Math.max(Math.floor(totalChars / 4), 1);
}

// ---------------------------------------------------------------------------
// Question routing
// ---------------------------------------------------------------------------

/**
 * Route questions from the collapsed manifest to the best-match batch in-place.
 *
 * For each question: first pass attempts a direct hit (any overlap between
 * `target_files` and a batch's `files`). On miss, falls back to the batch
 * with the most files sharing `from_file`'s directory. First batch wins on
 * tie (Python `overlap > best_overlap` — strict greater-than).
 */
function routeQuestions(
  questionIndex: Array<Record<string, unknown>>,
  batches: BatchDefinition[],
): void {
  if (batches.length === 0) return;

  for (const q of questionIndex) {
    const targetFiles = Array.isArray(q.target_files) ? (q.target_files as string[]) : [];
    const fromFile = typeof q.from_file === 'string' ? q.from_file : '';
    const qTextRaw = typeof q.text === 'string' ? q.text : q.question;
    const qText = typeof qTextRaw === 'string' ? qTextRaw : '';
    const qId = typeof q.id === 'string' ? q.id : '';

    const routedQuestion: RoutedQuestion = {
      id: qId,
      from_file: fromFile,
      question: qText,
    };

    // Direct match: any shared file between batch and question targets.
    const targetSet = new Set(targetFiles);
    let routed = false;
    for (const batch of batches) {
      const batchFiles = batch.files;
      let hit = false;
      for (const f of batchFiles) {
        if (targetSet.has(f)) {
          hit = true;
          break;
        }
      }
      if (hit) {
        batch.routed_questions.push(routedQuestion);
        routed = true;
        break;
      }
    }

    if (routed) continue;

    // Fallback: batch with most files in the same directory as from_file.
    const fromDir = pyDirname(fromFile);

    let bestBatch = batches[0] as BatchDefinition;
    let bestOverlap = -1;
    for (const batch of batches) {
      let overlap = 0;
      for (const f of batch.files) {
        if (pyDirname(f) === fromDir) overlap++;
      }
      if (overlap > bestOverlap) {
        bestOverlap = overlap;
        bestBatch = batch;
      }
    }
    bestBatch.routed_questions.push(routedQuestion);
  }
}

// ---------------------------------------------------------------------------
// Cluster helpers
// ---------------------------------------------------------------------------

/** Build a file-path -> cluster_id map from the structural map modules. */
function buildFileToCluster(structuralMap: StructuralMap): Map<string, number> {
  const mapping = new Map<string, number>();
  for (const module of structuralMap.modules ?? []) {
    mapping.set(module.path ?? '', module.cluster_id ?? 0);
  }
  return mapping;
}

/** Find a cluster by id; null when not present. */
function getClusterById(
  structuralMap: StructuralMap,
  clusterId: number,
): StructuralMapCluster | null {
  for (const cluster of structuralMap.clusters ?? []) {
    if (cluster.id === clusterId) return cluster;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Form Round 2 review batches from a collapsed manifest and structural map.
 *
 * Strategy:
 *   1. Group preserved cards by `cluster_id` from the structural map (files
 *      missing from the map fall back to cluster 0).
 *   2. Emit cluster groups in ascending cluster_id order. If a group exceeds
 *      `maxFilesPerBatch`, split alphabetically by top-level subdir, then
 *      slice each subdir group into `maxFilesPerBatch` chunks.
 *   3. Route `question_index` entries into batches (see {@link routeQuestions}).
 *   4. Attach cluster context (one entry per batch when the cluster exists).
 *
 * @returns Array of batch definitions conforming to BATCH_DEFINITION_SCHEMA.
 */
export function formBatches(
  collapsedManifest: CollapsedManifest,
  structuralMap: StructuralMap,
  maxFilesPerBatch = 15,
  projectRoot = '.',
): BatchDefinition[] {
  const preservedCards = Array.isArray(collapsedManifest.preserved_cards)
    ? collapsedManifest.preserved_cards
    : [];
  const questionIndex = Array.isArray(collapsedManifest.question_index)
    ? collapsedManifest.question_index
    : [];

  if (preservedCards.length === 0) return [];

  const fileToCluster = buildFileToCluster(structuralMap);

  // Group cards by cluster_id. Map insertion order matches first-seen; we
  // sort numerically before emission to mirror Python's sorted().
  const clusterGroups = new Map<number, TriageCard[]>();
  for (const card of preservedCards) {
    const filePath = typeof card.file === 'string' ? card.file : '';
    const clusterId = fileToCluster.get(filePath) ?? 0;
    const bucket = clusterGroups.get(clusterId);
    if (bucket) bucket.push(card);
    else clusterGroups.set(clusterId, [card]);
  }

  const rawBatches: BatchDefinition[] = [];
  let batchIdCounter = 1;

  const sortedClusterIds = [...clusterGroups.keys()].sort((a, b) => a - b);
  for (const clusterId of sortedClusterIds) {
    const cards = clusterGroups.get(clusterId) ?? [];
    const clusterCtx = getClusterById(structuralMap, clusterId);

    if (cards.length <= maxFilesPerBatch) {
      const files = cards.map((c) => (typeof c.file === 'string' ? c.file : ''));
      rawBatches.push({
        batch_id: batchIdCounter,
        files,
        triage_cards: cards,
        cluster_context: clusterCtx ? [clusterCtx] : [],
        routed_questions: [],
        estimated_tokens: _estimateTokens(files, projectRoot),
      });
      batchIdCounter++;
      continue;
    }

    // Split by top-level subdirectory alphabetically, then chunk.
    const subdirGroups = new Map<string, TriageCard[]>();
    for (const card of cards) {
      const filePath = typeof card.file === 'string' ? card.file : '';
      const subdir = pyDirname(filePath);
      const bucket = subdirGroups.get(subdir);
      if (bucket) bucket.push(card);
      else subdirGroups.set(subdir, [card]);
    }

    const sortedSubdirs = [...subdirGroups.keys()].sort();
    for (const subdir of sortedSubdirs) {
      const subCards = subdirGroups.get(subdir) ?? [];
      for (let i = 0; i < subCards.length; i += maxFilesPerBatch) {
        const chunk = subCards.slice(i, i + maxFilesPerBatch);
        const files = chunk.map((c) => (typeof c.file === 'string' ? c.file : ''));
        rawBatches.push({
          batch_id: batchIdCounter,
          files,
          triage_cards: chunk,
          cluster_context: clusterCtx ? [clusterCtx] : [],
          routed_questions: [],
          estimated_tokens: _estimateTokens(files, projectRoot),
        });
        batchIdCounter++;
      }
    }
  }

  routeQuestions(questionIndex, rawBatches);

  return rawBatches;
}
